File: lib/py/src/metrics.py
import os
import datetime
import tornado.gen
import statsd

if os.environ.get("STATSD_URL", None):
    ip, port = os.environ.get("STATSD_URL").split(':')
    statsd_client = statsd.StatsClient(host=ip, port=int(port),
                                       prefix=os.environ.get('STATSD_PREFIX',
                                                             None))
else:
    statsd_client = statsd.StatsClient()


def instrument(name):
    def instrument_wrapper(fn):
        @tornado.gen.coroutine
        def fn_wrapper(*args, **kwargs):
            def send_duration():
                duration = (datetime.datetime.now() - start).total_seconds()
                statsd_client.timing("{}.duration".format(name),
                                     int(duration * 1000))

            start = datetime.datetime.now()
            ftr_result = fn(*args, **kwargs)

            try:
                result = yield tornado.gen.maybe_future(ftr_result)
            except Exception as e:
                if statsd_client:
                    statsd_client.incr("{}.exceptions.{}".format(
                        name, e.__class__.__name__.lower()))
                    send_duration()

                raise e
            else:
                if statsd_client:
                    send_duration()
                    statsd_client.incr("{}.success".format(name))

                raise tornado.gen.Return(result)

        return fn_wrapper
    return instrument_wrapper

File: lib/rb/lib/thrift/metrics.rb
require 'statsd'

class NullStatsd
  def method_missing(*)
  end
end

module Thrift
  module Metrics
    module Middleware
      class Timing
        class << self
          def send_duration(name, t0)
            Metrics.client.timing(
              "#{name}.duration",
              ((Time.now - t0) * 1_000).to_i
            )
          end

          def instrument(name, &block)
            t0 = Time.now

            begin
              r = block.call
              send_duration name, t0
              r
            rescue => e
              send_duration name, t0
              raise e
            end
          end
        end
      end

      class Exception
        class << self
          def instrument(name, &block)
            begin
              r = block.call
              Metrics.client.increment("#{name}.success")
              r
            rescue => e
              Metrics.client.increment("#{name}.exceptions.#{e.class.name.downcase}")
              raise e
            end
          end
        end
      end
    end

    MIDDLEWARES = [Middleware::Timing, Middleware::Exception]

    class << self
      def instrument(name, &block)
        MIDDLEWARES.reduce(block) do |acc, cur|
          Proc.new { cur.instrument(name) { acc.call } }
        end.call
      end

      def client
        @client ||= begin
                      if ENV['STATSD_URL']
                        ip, port = ENV['STATSD_URL'].split(':')
                        Statsd.new ip, port.to_i
                      else
                        NullStatsd.new
                      end
                    end
      end
    end
  end
end
